echo-pubsub/src/main/java/com/netflix/spinnaker/echo/pubsub/amazon/SQSSubscriber.java

```diff
@@ -0,0 +1,267 @@
+      description.setArtifacts(parseArtifacts(description.getMessagePayload(), messageId));
+    }
+    pubsubMessageHandler.handleMessage(description, acknowledger, identity.getIdentity(), messageId);
+  }
+
+  private List<Artifact> parseArtifacts(String messagePayload, String messageId){
+    List<Artifact> artifacts = messageArtifactTranslator.parseArtifacts(messagePayload);
+    if (artifacts == null) {
+      log.debug("No artifacts were found for subscription: {} and messageId: {}", subscription.getName(), messageId);
+    } else {
+      artifacts.forEach( a -> a.setArtifactAccount(subscription.getAccountName()));
+      log.debug(
+        "Artifacts found for subscription: {}: {}",
+        subscription.getName(),
+        String.join(", ", artifacts.stream().map(Artifact::toString).collect(Collectors.toList()))
+      );
+    }
+    return artifacts;
+  }
+
+  private String unmarshallMessageBody(String messageBody) {
+    String messagePayload = messageBody;
+    try {
+      NotificationMessageWrapper wrapper = objectMapper.readValue(messagePayload, NotificationMessageWrapper.class);
+      if (wrapper != null && wrapper.getMessage() != null) {
+        messagePayload = wrapper.getMessage();
+      }
+    } catch (IOException e) {
+      // Try to unwrap a notification message; if that doesn't work,
+      // we're dealing with a message we can't parse. The template or
+      // the pipeline potentially knows how to deal with it.
+      log.error("Unable unmarshal NotificationMessageWrapper. Unknown message type. (body: {})", messageBody, e);
+      registry.counter(getNotAmazonNotificationMetricId(subscriptionName())).increment();
+    }
+    return messagePayload;
+  }
+
+
+
+  private static String ensureQueueExists(AmazonSQS amazonSQS,
+                                          ARN queueARN,
+                                          ARN topicARN,
+                                          int sqsMessageRetentionPeriodSeconds) {
+    String queueUrl = amazonSQS.createQueue(queueARN.getName()).getQueueUrl();
+    log.debug("Created queue " + queueUrl);
+
+    HashMap<String, String> attributes = new HashMap<>();
+    attributes.put("Policy", buildSQSPolicy(queueARN, topicARN).toJson());
+    attributes.put("MessageRetentionPeriod", Integer.toString(sqsMessageRetentionPeriodSeconds));
+    amazonSQS.setQueueAttributes(
+      queueUrl,
+      attributes
+    );
+
+    return queueUrl;
+  }
+
+  private static String subscribeToTopic(AmazonSNS amazonSNS,
+                                         ARN topicARN,
+                                         ARN queueARN) {
+    amazonSNS.subscribe(topicARN.getArn(), "sqs", queueARN.getArn());
+    return topicARN.getArn();
+  }
+
+  /**
+   * This policy allows operators to choose whether or not to have pubsub messages to be sent via SNS for fanout, or
+   * be sent directly to an SQS queue from the autoscaling group.
+   */
+  private static Policy buildSQSPolicy(ARN queue, ARN topic) {
```

Comment:
a few of these methods were lifted from `clouddriver`?

might be worth also extracting them vs. duplicating?

Reply:
I've put a todo in to extract to `kork-aws`.